src/cart/crt_hlc.c

```diff
@@ -0,0 +1,117 @@
+/* Copyright (C) 2019 Intel Corporation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted for any purpose (including commercial purposes)
+ * provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions, and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions, and the following disclaimer in the
+ *    documentation and/or materials provided with the distribution.
+ *
+ * 3. In addition, redistributions of modified forms of the source or binary
+ *    code must carry prominent notices stating that the original code was
+ *    changed and the date of the change.
+ *
+ * 4. All publications or advertising materials mentioning features or use of
+ *    this software are asked, but not required, to acknowledge that it was
+ *    developed by Intel Corporation and credit the contributors.
+ *
+ * 5. Neither the name of Intel Corporation, nor the name of any Contributor
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+/**
+ * This file is part of CaRT. Hybrid Logical Clock (HLC) implementation.
+ */
+#include "crt_hlc.h"
+#include <gurt/common.h>	/* for NSEC_PER_SEC */
+#include <gurt/atomic.h>
+#include <time.h>
+
+#if defined(HAVE_STDATOMIC)
+#if defined(__INTEL_COMPILER)
+#pragma warning(push)
+/**
+ * argument of type "_Atomic(uint64_t={__uint64_t={unsigned long}}) *" is
+ * incompatible with parameter of type "const volatile void *"
+ * (dropping qualifiers)
+ */
+#pragma warning(disable:2330)
+#endif /* __INTEL_COMPILER */
+#endif /* HAVE_STDATOMIC */
+
+#define CRT_HLC_MASK 0xFFFFULL
+
+static ATOMIC uint64_t crt_hlc;
+
+/** Get local physical time */
+static inline uint64_t crt_hlc_localtime_get(void)
+{
+	struct timespec now;
+	uint64_t	pt;
+	int		rc;
+
+	rc = clock_gettime(CLOCK_REALTIME_COARSE, &now);
+	pt = rc ? atomic_load_consume(&crt_hlc)
```

Comment:
I observed few fails in my local testing. This error affected correct ordering. So, I resolved this. In case of error I just use previous time.

Reply:
That seems error prone.   What if previous time is a long time ago?

Reply:
Based on man page, the only two errors could be returned are, EINVAL and EPERM, the first one is about invalid parameters, so I'm wondering how it could fail in your case

Reply:
I didn't remember exact error. Probably it was EBUSY. Anyway if I use old time value it works stable.